// @8base/table-schema-provider/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TableSchemaContext", {
  enumerable: true,
  get: function get() {
    return _TableSchemaContext.TableSchemaContext;
  }
});
Object.defineProperty(exports, "TABLES_SCHEMA_QUERY", {
  enumerable: true,
  get: function get() {
    return _TableSchemaProvider.TABLES_SCHEMA_QUERY;
  }
});
Object.defineProperty(exports, "TABLE_FRAGMENT", {
  enumerable: true,
  get: function get() {
    return _TableSchemaProvider.TABLE_FRAGMENT;
  }
});
Object.defineProperty(exports, "TABLE_FIELD_FRAGMENT", {
  enumerable: true,
  get: function get() {
    return _TableSchemaProvider.TABLE_FIELD_FRAGMENT;
  }
});
Object.defineProperty(exports, "TableSchemaProvider", {
  enumerable: true,
  get: function get() {
    return _TableSchemaProvider.TableSchemaProvider;
  }
});
Object.defineProperty(exports, "TableConsumer", {
  enumerable: true,
  get: function get() {
    return _TableConsumer.TableConsumer;
  }
});

var _TableSchemaContext = require("./TableSchemaContext");

var _TableSchemaProvider = require("./TableSchemaProvider");

var _TableConsumer = require("./TableConsumer");
