// @8base/utils/dist/selectors/tableFieldSelectors.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRelationTableName = exports.getRelationTableId = exports.getFieldName = exports.isListField = exports.isMetaField = exports.isIdField = exports.isSmartField = exports.isFileField = exports.isRelationField = exports.getFieldTypesAttributes = exports.getFieldType = exports.getTableField = void 0;

var R = _interopRequireWildcard(require("ramda"));

var _reselect = require("reselect");

var _constants = require("../constants");

var getTableField = function getTableField(tableField) {
  return tableField;
};

exports.getTableField = getTableField;
var getFieldType = (0, _reselect.createSelector)(getTableField, R.prop('fieldType'));
exports.getFieldType = getFieldType;
var getFieldTypesAttributes = (0, _reselect.createSelector)(getTableField, R.prop('fieldTypeAttributes'));
exports.getFieldTypesAttributes = getFieldTypesAttributes;
var isRelationField = (0, _reselect.createSelector)(getFieldType, R.equals(_constants.FIELD_TYPE.RELATION));
exports.isRelationField = isRelationField;
var isFileField = (0, _reselect.createSelector)(getFieldType, R.equals(_constants.FIELD_TYPE.FILE));
exports.isFileField = isFileField;
var isSmartField = (0, _reselect.createSelector)(getFieldType, R.equals(_constants.FIELD_TYPE.SMART));
exports.isSmartField = isSmartField;
var isIdField = (0, _reselect.createSelector)(getFieldType, R.equals(_constants.FIELD_TYPE.ID));
exports.isIdField = isIdField;
var isMetaField = (0, _reselect.createSelector)(getTableField, R.propEq('isMeta', true));
exports.isMetaField = isMetaField;
var isListField = (0, _reselect.createSelector)(getTableField, R.propEq('isList', true));
exports.isListField = isListField;
var getFieldName = (0, _reselect.createSelector)(getTableField, R.prop('name'));
exports.getFieldName = getFieldName;
var getRelationTableId = (0, _reselect.createSelector)(getTableField, R.path(['relation', 'refTable', 'id']));
exports.getRelationTableId = getRelationTableId;
var getRelationTableName = (0, _reselect.createSelector)(getTableField, R.path(['relation', 'refTable', 'name']));
exports.getRelationTableName = getRelationTableName;
